[Feature] follow 추천 기능 추가

dev-msj opened the pull request:
## 작업 내용

유저 추천 목록 30명 가져오기
- 관심 카테고리가 1개 이상 겹치면서, 어떤 관계도 맺지 않은 사람 30명을 추천
- 공통된 관심 카테고리를 제외한 나머지 카테고리의 개수가 많으면서, 작성한 글이 많은 순으로 정렬
- 위 조건에 해당하는 사람의 수가 30명보다 적을 경우, 작성글이 많은 순으로 부족한만큼 채우기
- 관심 카테고리가 없거나 관심 카테고리가 겹치는 사람이 없을 경우, 작성글이 많은 순으로 30명 추천
- 한 번 카테고리를 가져오면 3분 동안 캐싱

UserRelation 추가하기
- relation을 생성할 relationUser, relationType(FOLLOW/BLACK_LIST)를 입력 받아 USER_RELATION 테이블에 추가
- user relation이 생성되면 redis에서 해당 유저 cache 삭제(CacheEvict)

## 이슈 번호

#37

dev-msj committed:
docs : 회원가입 기능 정의

dev-msj committed:
feat(UserAuthMapper) : 신규 UserAuth insert 기능 추가

dev-msj committed:
feat(UserAuthRepository) : creatUserAuth 메서드 추가

dev-msj committed:
refactor(UserAuth) : 외부 통신에 사용될 수 있도록 수정

AllArgsConstructor, NoArgsConstructor 추가
Serializable 상속

dev-msj committed:
feat(UserAuthService) : createUserAuth 메서드 추가

salt 생성, password hashing 후 UserAuth Insert

dev-msj committed:
feat(UserAuthController) : createUserAuth 메서드 추가

dev-msj committed:
Merge branch 'feature/feed-write' into feature/user-join

# Conflicts:
#	src/main/java/com/flab/oasis/service/UserAuthService.java

dev-msj committed:
refactor(UserAuthService) : createUserAuth에서 인증된 uid를 사용하도록 수정

dev-msj committed:
Merge branch 'fix/login-google/login-handler' into feature/user-join

# Conflicts:
#	src/main/java/com/flab/oasis/service/UserAuthService.java

dev-msj committed:
fix(UserAuthMapper) : 파라메터명 오류 수정

dev-msj committed:
Merge branch 'main' into feature/user-join

# Conflicts:
#	src/main/java/com/flab/oasis/controller/UserAuthController.java
#	src/main/java/com/flab/oasis/model/UserAuth.java
#	src/main/java/com/flab/oasis/service/UserAuthService.java

dev-msj committed:
Merge branch 'main' into feature/user-join

dev-msj committed:
fix(UserAuthService) : merge된 내용 반영

dev-msj committed:
Merge branch 'main' into feature/user-join

dev-msj committed:
refactor(UserAuthService) : createUserAuth 로직 수정

UserRole을 'USER'로 설정
socialYN이 'N'일 때 비밀번호를 hashing하도록 변경
LoginResult를 반환하도록 수정

dev-msj committed:
refacot(UserAuthController) : 회원가입 시 LoginResult를 반환하도록 수정

dev-msj committed:
docs : 회원가입 절차 재정의

dev-msj committed:
feat : UserInfo 모델 추가

dev-msj committed:
fix(UserAuthMapper) : tag명 오류 수정

insert 구문의 tag명 수정 select -> insert

dev-msj committed:
refactor(UserAuthMapper) : UserAuth 생성시 UserRole도 생성하도록 수정

dev-msj committed:
refactor(UserCategory) : Builder 패턴 추가

dev-msj committed:
feat(UserCategoryMapper) : 카테고리 생성 기능 추가

dev-msj committed:
feat : UserProfile 모델 추가

UserInfo와 UserCategory를 통합하여 client와 통신하는 객체

dev-msj committed:
feat(UserInfoMapper) : 닉네임 중복 확인 쿼리 추가

dev-msj committed:
feat : 닉네임 중복확인 기능 추가

dev-msj committed:
refactor(MyBatisConfigUser) : transaction manager 추가

dev-msj committed:
feat(UserInfoMapper) : USER_INFO 생성 및 조회 쿼리 추가

dev-msj committed:
feat : NotFoundException 추가

dev-msj committed:
feat(UserInfoRepository) : UserInfo 생성 및 조회 기능 추가

UserInfo 조회 시 redis에서 먼저 조회.
조회 결과가 null이면 DB에서 가져온 후 redis에 저장.

DB에서 조회 시 null일 경우 NoFoundException을 던지도록 함.
-> NotFoundException일 경우 client에서 프로필 생성 화면을 보여주게 하기 위함

dev-msj committed:
feat(UserCategoryRepository) : UserCategory 생성 및 조회 기능 추가

UserCategory 조회 시 redis에서 먼저 조회.
조회 결과가 null이면 DB에서 가져온 후 redis에 저장.

dev-msj committed:
feat(UserProfileService) : User Profile을 생성 및 조회하는 기능 추가

User Profile을 생성할 때 Transactional 선언으로 에러 시 rollback되도록 처리

dev-msj committed:
feat(UserInfoController) : User Profile 생성 및 조회 API 추가

dev-msj committed:
feat(ExceptionAdvice) : NotFoundException handler 추가

dev-msj committed:
fix(UserCategoryMapper) : UserCategoryList를 foreach 구문으로 처리

dev-msj committed:
fix(UserProfileService) : UserCategory 생성 관련 기능 오류 수정

uid를 SecurityContext에서 받아오도록 수정
UserCategory가 empty가 아닐 경우만 insert하도록 수정

dev-msj committed:
refactor(BookCategory) : Getter를 Lombok 어노테이션으로 대체

dev-msj committed:
refactor(HomeService) : UserCategoryRepository 반영

기존 UserCategoryMapper를 UserCategoryRepository로 대체

dev-msj committed:
Merge branch 'feature/user-join'

dev-msj committed:
refactor(UserCategoryRepository) : 무의미한 코드 수정

기본적으로 List형을 반환하므로 Nullable로 List 객체를 생성하는 코드 제거

dev-msj committed:
style : 유저 카테고리를 가져오는 메서드명 수정

UserCategory List를 반환하므로 반환형에 맞게 메서드명 수정
getUserCategoryByUid -> getUserCategoryListByUid

dev-msj committed:
feat(SQLResultResponse) : SQL 실행 결과를 반환하는 객체 추가

dev-msj committed:
refactor : 프로필 생성 관련 쿼리 excpetion 처리

spring은 Runtime 중 발생하는 모든 SQLException을 DataAccessException으로 포장해서 던진다.
그러므로 Repository에서 DataAccessException으로 catch하여 다시 SQLException를 throw하도록 변경

dev-msj committed:
refactor : 프로필 생성 결과를 SQLResultResponse로 반환하도록 수정

dev-msj committed:
feat(ResultResponse) : API 응답용 공통 모델 추가

Generic을 통해 응답할 data의 클래스를 추상화

dev-msj committed:
refactor : UserInfo/Category repository에서 try-catch 삭제

dev-msj committed:
refactor(UserProfileService) : ExceptionAdvice가 handling하도록 수정

SQLException은 모든 Service의 공통 관심사이므로 ExceptionAdvice가 처리하도록 수정

dev-msj committed:
refactor(UserProfileService) : ResultResponse를 반환하도록 수정

dev-msj committed:
refctor(UserProfileService) : NotFoundException을 catch하도록 수정

dev-msj committed:
refactor(UserProfileController) : ResultResponse를 응답하도록 수정

dev-msj committed:
refactor(AuthenticationException) : field에 value 추가

dev-msj committed:
style : 사용되지 않는 모델 삭제

dev-msj committed:
refactor(ExceptionAdvice) : ResultResponse로 응답하도록 수정

dev-msj committed:
refactor(JwtFilter) : 토큰 갱신 응답을 ResultResponse로 대체

토큰 갱신 응답을 ResponseReissuedJWT를 ResultResponse로 대체
ResponseReissuedJWT 삭제

dev-msj committed:
refactor(UserProfileService) : 정상처리 시 code를 0으로 반환하도록 수정

dev-msj committed:
refactor : API Response 모델명 변경

모델명 변경 ResultResponse -> GeneralResponse

dev-msj committed:
docs(UserAuthService) : 유저 생성 로그 추가

dev-msj committed:
refactor : GeneralResponse를 Controller가 생성하도록 수정

dev-msj committed:
fix(UserInfoRepository) : redis 관련 로직 에러 수정

redis에 값이 없을 경우에만 db에서 가져온 값을 redis에 등록하도록 변경
redis의 get 결과를 직접 optional로 감쌀 경우 null로 처리되는 문제 해결
json convertValue 시 TypeReference가 아닌 class 정보를 직접 전달하도록 변경

dev-msj committed:
fix(UserInfoRepository) : redis 관련 exception 처리 변경

redis에 연결되지 않을 경우에만 DB를 조회하도록 Exception 변경
Exception -> RedisConnectionFailureException

dev-msj committed:
docs(UserInfoRepository) : redis 연결 실패 로그 추가

dev-msj committed:
refactor(ExceptionAdvice) : SQLException을 DataAccessException으로 대체

runtime 중에 발생한 jdbc에러를 DataAccessException로 감싸서 던지기 때문에 advice 대상을 변경

dev-msj committed:
Merge branch 'main' into feature/user-join

dev-msj committed:
style(ExceptionAdvice) : 사용되지 않는 import 제거

dev-msj committed:
Merge branch 'feature/user-join' of https://github.com/dev-msj/Oasis into feature/user-join

dev-msj committed:
feat(ResponseCode) : 정상 응답용 코드 추가

dev-msj committed:
style(ErrorCode) : RESET_CONTENT 삭제

ResponseCode로 대체

dev-msj committed:
refactor(JwtFilter) : ResponseCode의 RESET_CONTENT로 대체

dev-msj committed:
refactor(UserProfileController) : 응답 코드에 ResponseCode 적용

dev-msj committed:
feat : BaseResponse 추가

dev-msj committed:
feat : SuccessResponse 추가

dev-msj committed:
refactor(UserProfileController) : isExistsNickname의 반환형을 boolean으로 변경

dev-msj committed:
refactor(UserProfileController) : createUserProfile의 결과로 SuccessResponse 반환

dev-msj committed:
feat : UserProfileResponse 추가

dev-msj committed:
refactor(UserProfileController) : getUserProfileByUid를 반환형 변경

GeneralResponse에서 UserProfileResponse로 대체

dev-msj committed:
refactor(ExceptionAdvice) : DataAccessException handler 삭제

dev-msj committed:
feat : FailureResponse 추가

dev-msj committed:
refactor(ExceptionAdvice) : 응답을 FailureResponse로 대체

dev-msj committed:
refactor(ExceptionAdvice) : exception이 발생한 value를 알려주지 않도록 변경

dev-msj committed:
feat : JsonWebTokenResponse 추가

dev-msj committed:
refactor(JwtFilter) : 토큰 갱신용 응답 모델 변경

GeneralResponse -> JsonWebTokenResponse

dev-msj committed:
style(GeneralResponse) : 사용되지 않는 모델 삭제

dev-msj committed:
style(UserProfileService) : 사용되지 않는 import 삭제

dev-msj committed:
fix(JwtServiceTest) : assert Exception 오류 해결

AuthorizationException을 삭제하고 모두 AuthenticationException으로 대체하였으나
미반영된 내용 해결

dev-msj committed:
refactor(FeedController) : SuccessResponse 적용

dev-msj committed:
feat : RedisErrorHandler 추가

Redis에 문제가 발생할 경우 서비스 오류가 발생하지 않고, Log 생성 후 DB의 값을 전달

dev-msj committed:
refactor(UserInfoRepository) : redisTemplate을 Cacheable로 대체

dev-msj committed:
style(RedisConfig) : 메서드에서 static 제거

dev-msj committed:
style(UserLoginRequest) : 불필요한 상속 제거

dev-msj committed:
fix(RedisConfig) : ClassCastException 문제 해결

redis의 data를 deserialize할 때 LinkedHashMap이 원하는 Model로 cast되지 않는 문제 해결
- 원인 : LinkedHashMap은 bootstrap loader에 의해 올라가고, Model은 app loader에 의해 올라가기 때문에
변환될 수 없음.
- 해결 : 직접 선언했던 ObjectMapper를 삭제하고, GenericJackson2JsonRedisSerializer에서 기본으로 생성하는
ObjectMapper를 사용하게 함으로써 해결됨.

dev-msj committed:
refactor : cacheable 어노테이션 적용

dev-msj committed:
refactor : Exception 로직 이전

UserProfileController
- getUserProfileByUid 메서드에서 NotFoundException catch 로직 삭제

UserProfileService
- getUserProfileByUid 메서드에 NotFoundException catch 로직 추가

dev-msj committed:
style(BaseResponse) : 중복된 implements 제거

dev-msj committed:
refactor : 회원가입 시 JWT만 반환하도록 변경

dev-msj committed:
feat(ErrorCode) : EXPECTATION_FAILED 코드 추가

dev-msj committed:
feat : UnexpectedCodeException 추가

dev-msj committed:
refactor(FailureResponse) : code 검증 로직 추가

hiddenBuilder를 통해 code가 200일 경우 UnexpectationCodeException을 던지도록 변경

dev-msj committed:
style : 사용하지 않는 import 제거

dev-msj committed:
refactor(TestMapper) : 범용 메서드를 사용하도록 변경

Select 어노테이션을 활용한 getTestData 메서드 추가
불필요한 Repository 어노테이션 삭제
cacheTest 메서드 삭제
dbConnectionTest 메서드 삭제

dev-msj committed:
style(TestMapper.xml) : 사용되지 않는 파일 삭제

dev-msj committed:
refactor(TestService) : Cache 기능만 Test하도록 변경

dbConnectionTest 메서드 삭제
EhCache/Redis의 Caching 및 Evict Test 메서드 추가
불필요한 import 삭제

dev-msj committed:
test(CommonTest) : MyBatis/Caching Test 재작성

MyBatis Connection Test
- 불필요하게 TestModel을 사용하지 않고 직접 데이터를 받아와 assert하도록 변경

Redis/EhCache Caching Test
- Mokito 적용
- 동일 메서드를 중복 호출하여 캐싱에 의해 실제 동작 횟수가 1회인지 확인

Redis/EhCache Cache Evict Test
- Mokito 적용
- 특정 메서드에 대한 캐시를 삭제한 후에 해당 메서드가 재호출되는지 확인

dev-msj committed:
style : 사용되지 않는 TestModel 삭제

dev-msj committed:
Merge branch 'feature/user-join'

dev-msj committed:
docs : 팔로잉 추천 추가

dev-msj committed:
feat : RelationType 추가

dev-msj committed:
feat : UserRelation 추가

dev-msj committed:
fix(UserRelation) : 누락된 컬럼 추가

dev-msj committed:
feat : UserRelationMapper 추가

dev-msj committed:
feat : OverlappingCategoryUserSelect 추가

getUidListWithOverlappingBookCategory 쿼리에 필요한 field를 담은 모델 생성

dev-msj committed:
feat(UserCategoryMapper) : getUidListWithOverlappingBookCategory 쿼리 추가

dev-msj committed:
feat(UserCategoryRepository) : getUidListWithOverlappingBookCategory 메서드 추가

dev-msj committed:
docs : 오타 수정

dev-msj committed:
feat : UserCategoryCountSelect 추가

dev-msj committed:
feat : UserCategoryCount 추가

dev-msj committed:
feat(UserCategoryMapper) : getUserCategoryCountList 쿼리 추가

dev-msj committed:
feat(UserCategoryRepository) : getUserCategoryCountList 메서드 추가

dev-msj committed:
feat : FeedCountSelect 추가

dev-msj committed:
feat : FeedCount 추가

dev-msj committed:
feat(FeedMapper) : FeedCountList 쿼리 추가

dev-msj committed:
feat : UserRelationRepository 추가

getUserRelationListByUid 메서드 추가
Caceable을 통해 redis에 UserRelationList을 저장

dev-msj committed:
feat : RecommendUser 추가

Comparable을 상속하여 카테고리의 개수가 많은 순, 피드가 많은 순으로 정렬

dev-msj committed:
feat : UserRelationService 추가

getRecommendUserList 메서드 구현

dev-msj committed:
feat(UserRelationMapper) : UserRelation 생성 쿼리 추가

dev-msj committed:
feat(UserRelationRepository) : UserRelation 생성 메서드 추가

CacheEvict를 통해 대상 uid를 key로 가진 cache 삭제

dev-msj committed:
feat(UserRelationService) : UserRelation 생성 메서드 추가

dev-msj committed:
feat : UserRelationController 추가

getRecommendUserList API 추가
createUserRelation API 추가

dev-msj committed:
fix(FeedMapper) : 조건문 오류 해결

otherwise에서 '본인/relation이 있는 user/생성된 recommend user'를 제외하도록 수정

dev-msj committed:
refactor(FeedCountSelect) : 사용하지 않는 field 삭제

dev-msj committed:
feat(ehcache) : recommendUserCache 추가

dev-msj committed:
refactor(UserRelationService) : getRecommendUserList 로직 개선

getRecommendUserList의 cache key를 위해 uid를 파라메터로 넘기도록 변경
Cacheable을 통해 getRecommendUserList가 3분간 캐싱되도록 수정
getRecommendUserList 로직 정리

dev-msj committed:
refactor(UserRelationController) : UserRelationController에 uid를 넘기도록 수정

dev-msj committed:
fix(UserCategoryMapper) : getUidListWithOverlappingBookCategory 쿼리 오류 해결

userRelationList가 empty가 아닐 때만 사용하도록 변경

dev-msj committed:
fix(FeedMapper) : WHERE 조건문 삭제

작성된 피드가 없으면 Feed에서 추천 유저를 가져오는게 무의미하므로 feed count만 하도록 변경
관련 모델인 FeedCountSelect 삭제

dev-msj committed:
style(FeedMapper) : 사용되지 않는 import 삭제

dev-msj committed:
feat : RecommendUserSelect 추가

dev-msj committed:
feat : getRecommendUserList 쿼리 추가

추천 유저를 Feed Count 순으로 needSize만큼 가져온다.
Feed는 nullable하므로 join하지 않고 subquery로 사용한다.

dev-msj committed:
feat(UserInfoRepository) : getRecommendUserList 메서드 추가

dev-msj committed:
refactor(UserRelationService) : getRecommendUserAsManyAsNeed 메서드 관련 로직 수정

getFeedCountList 변경에 맞게 수정
UserInfoRepository의 getRecommendUserList 적용

dev-msj committed:
fix(UseCategoryMapper) : getUserCategoryCountList의 resultType 에러 해결

dev-msj committed:
fix(FeedMapper) : getFeedCountList의 조건문 오류 해결

dev-msj committed:
fix(FeedMapper) : getFeedCountList 쿼리 삭제

Feed가 존재하지 않으면 uid별 feed 개수 값을 가져오는 것이 무의미하므로 삭제
해당 쿼리와 관련된 FeedCount 모델 삭제

dev-msj committed:
feat : UsersFeedCountSelect 모델 추가

dev-msj committed:
feat : UsersFeedCount 모델 추가

dev-msj committed:
style : 미사용 중인 RecommendUserSelect 삭제

dev-msj committed:
refactor(UserInfoMapper) : 메서드명 및 조건문 추가

메서드명 변경
- getRecommendUserList -> getUsersFeedCountList

needSize 값에 따른 조건문 추가
- '-1'일 경우에는 원하는 user들의 uid와 feed count 가져오기
- '-1'가 아닐 경우에는 원하지 않는 user들을 제외하고 uid와 feed count를 needSize만큼 가져오기

dev-msj committed:
refactor(UserInfoRepository) : 메서드 수정

메서드명 변경
- getRecommendUserList -> getUsersFeedCountList

파라메터 변경
- RecommendUserSelect -> UsersFeedCountSelect

리턴값 변경
- List<String> -> List<UsersFeedCount>

dev-msj committed:
refactor(UserRelationService) : FeedMapper를 UserInfoRepository로 대체

FeedCount 값을 UserInfo를 통해 가져오도록 수정
- uid별 feed count를 가져와야하는데, feed 값이 없으면 feed에서는 select 할 수 없다.
  그래서 UserInfo 테이블에서 feed count를 subquery로 가져오는 방식으로 대체함.

dev-msj committed:
refactor : 모델에 Builder 패턴 적용

UserCategoryCount
UsersFeedCount

dev-msj committed:
refactor : 모델 및 메서드명 패턴 통일

UsersFeed{...} 형태가 아닌 UserFeed{...} 형태로 변경

dev-msj committed:
test : UserRelationServiceTest 추가

getRecommendUserList 메서드에 대한 테스트 코드 3가지 케이스 작성
- User Category가 존재하지 않을 경우
- 겹치는 User Category가 존재하지 않을 경우
- 생성된 추천 유저가 30명이 안될 경우

jinspark-lab commented on src/main/java/com/flab/oasis/model/UserCategoryCountSelect.java:
```diff
@@ -0,0 +1,18 @@
+    private static final long serialVersionUID = 1389467238782839862L;
+
+    private List<String> uidList;
+    private List<UserCategory> userCategoryList;
```
이 모델 구조가 잘 이해가 되지 않습니다. 설명 부탁드립니다.

dev-msj replied:
uidList에는 카테고리가 겹치는 유저들의 uid(overlappingCategoryUserList)를 담습니다.
userCategoryList에는 서비스를 요청한 유저의 category list를 담습니다.

이를 통해 getUserCategoryCountList 쿼리에서 
카테고리가 겹치는 유저들의 category에서 서비스를 요청한 유저의 category를 제외한 카테고리의 개수를 count하기 위해 사용하게 됩니다.

단순히 uidList로 사용하는 것보단 overlappingCategoryUidList로 변경하는 것이 더 좋을 것 같네요.

jinspark-lab commented on src/main/java/com/flab/oasis/repository/UserRelationRepository.java:
```diff
@@ -0,0 +1,26 @@
+
+    @CacheEvict(cacheNames = "UserRelation", key = "#userRelation.uid", cacheManager = "redisCacheManager")
+    public void createUserRelation(UserRelation userRelation) {
+        userRelationMapper.createUserRelation(userRelation);
```
정확히는 addRelation 으로 동작하는 구조인거죠?

dev-msj replied:
네 맞습니다.

jinspark-lab replied:
여기서 Evict 는 왜 빼신건가요?

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserRelationService.java:
```diff
@@ -0,0 +1,121 @@
+    private final UserRelationRepository userRelationRepository;
+    private final UserInfoRepository userInfoRepository;
+
+    @Cacheable(cacheNames = "recommendUserCache", key = "#uid", cacheManager = "ehCacheCacheManager")
```
ehCache 로 쓰시기는 좋지 않아보입니다. 이유에 대해서는 생각해보시는 걸 추천드립니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserRelationService.java:
```diff
@@ -0,0 +1,121 @@
+            recommendUserList.addAll(
+                    getRecommendUserAsManyAsNeed(uidList, 30 - recommendUserList.size())
+            );
+        }
```
이렇게 다루는것보다 limit 을 이용해서 좀 더 효과적으로 다루실 수 있습니다.
전체적으로 30명 관련 로직 리팩토링 가능합니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserRelationService.java:
```diff
@@ -0,0 +1,121 @@
+        List<String> recommendUserList = makeRecommendUserList(userCategoryList, overlappingCategoryUserList);
+
+        // 추천 유저가 30명이 안될 경우, 부족한 추천 user를 채운다.
+        if (recommendUserList.size() < 30) {
```
30을 매번 하드코딩하는 것보다 상수로 관리하시는게 낫습니다.

jinspark-lab replied:
변경 가능해야한다면 어떻게 하는게 좋을지도 고민해보세요

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserRelationService.java:
```diff
@@ -0,0 +1,121 @@
+        userRelationRepository.createUserRelation(userRelation);
+    }
+
+    // '본인 | relation이 있는 user | 생성된 recommend user'를 제외하고 needSize만큼 recommend user를 가져온다.
```
이 부분이 어디에 구현되어있는건가요? 설명 부탁드립니다.

dev-msj replied:
userCategoryList 또는 overlappingCategoryUserList가 Empty일 경우에는 userRelationList를 넘기도록 되어 있습니다.
그리고 recommendUserList가 30명이 안될 경우에는 userRelationList와 recommendUserList를 합쳐서 넘기도록 했습니다.
마지막으로 getRecommendUserAsManyAsNeed 메서드 안에서 getAuthenticatedUid를 통해 uid를 uidList에 추가합니다.

getRecommendUserAsManyAsNeed 메서드의 파라메터명을 uidList라고 하는 것보다 excludeUidList라고 하는 것이 의미 전달에 더 좋을 것 같네요.

jinspark-lab commented on src/test/java/com/flab/oasis/user/UserRelationServiceTest.java:
```diff
@@ -0,0 +1,159 @@
+                                .build()
+                ));
+
+        List<String> uidList = userRelationService.getRecommendUserList(uid);
```
로직이 어떻게 동작하는지 잘 모르겠습니다. 설명 부탁드립니다.

dev-msj replied:
이 테스트 관심 대상은 
1. UserCategory가 존재하지 않을 경우, 바로 getRecommendUserAsManyAsNeed 메서드를 call하는지
2. getUserFeedCountList 쿼리를 호출하여 최대 30명의 추천 유저를 가져와서 해당 uidList를 정상적으로 반환하는지
입니다.

그래서 
1. getUserCategoryListByUid 쿼리의 결과로 empty ArrayList를 반환하게 하고<br>`given(userCategoryRepository.getUserCategoryListByUid(uid)).willReturn(new ArrayList<>())`
2. getUserFeedCountList 쿼리의 결과로 expectedUid를 가진 List를 반환하게 하였습니다.<br>`given(userInfoRepository.getUserFeedCountList(BDDMockito.any(UserFeedCountSelect.class)))
                .willReturn(Collections.singletonList(UserFeedCount.builder().uid(expectedUid).feedCount(0).build()))`

마지막으로 `userRelationService.getRecommendUserList(uid)`의 실제 결과값에 expectedUid가 존재하는지 assertion하도록 했습니다.

jinspark-lab reviewed (changes requested):
추천 로직쪽, Counting 하는 부분, Select 관련 로직이 이해가 잘 되지 않습니다. 1차 피드백드립니다.

dev-msj committed:
refactor(UserCategoryCountSelect) : field명 변경

uidList는 실제 담아가는 값이 카테고리가 겹치는 유저들의 uid들이므로
field명을 overlappingCategoryUserList로 변경한다.

해당 모델을 사용하는 관련 코드들 수정

dev-msj committed:
refactor(UserRelationService) : getRecommendUserAsManyAsNeed의 parameter명 수정

uidList의 역할은 getUserFeedCountList 쿼리를 진행할 때 select 대상에서 제외하는 uid를 담는다.
해당 역할에 맞게 parameter명을 excludeUidList로 변경한다.

해당 메서드를 사용하는 관련 코드에도 해당 내용 반영
주석문 추가

jinspark-lab commented on src/main/resources/mapper/user/UserCategoryMapper.xml:
```diff
@@ -3,10 +3,52 @@
+                <foreach collection="userCategoryList" item="item" separator=",">
+                    #{item.bookCategory}
+                </foreach>
+            )
```
어플리케이션에서 처리하는걸 추천드립니다.

jinspark-lab reviewed (commented):
수업시간에 피드백드렸습니다. 리팩토링 한번 부탁드려요

dev-msj committed:
refactor(UserCategoryMapper) : getUidListWithOverlappingBookCategory 쿼리 수정

join을 활용해 겹치는 카테고리가 있는 유저 목록을 가져오도록 수정
관계가 존재하는 유저는 java 로직을 활용하여 필터링 하도록 수정

dev-msj committed:
refactor(UserCategoryRepository) : UserCategoryMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : UserCategoryMapper 변경사항 반영

카테고리가 겹치는 유저를 가져온 후 relation이 존재하는 유저를 필터링하도록 변경

dev-msj committed:
test(UserRelationServiceTest) : UserCategoryMapper 변경사항 반영

dev-msj committed:
style(OverlappingCategoryUserSelect) : 사용되지 않는 모델 삭제

dev-msj committed:
style(UserRelationService) : 사용되지 않는 메서드 삭제

dev-msj committed:
refactor(UserCategoryMapper) : getUserCategoryCountList 쿼리 수정

메서드명 변경
- getUserCategoryCountList -> getOverlappingUserCategoryList

쿼리 변경
- parameter로 전달받은 overlappingCategoryUserList와 user_category를 join하여
  overlappingCategoryUserList에 해당하는 UserCategory List를 반환하도록 수정

dev-msj committed:
refactor : UserCategoryMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : UserCategoryMapper 변경사항 반영

overlappingCategoryUserList에 해당하는 UserCategory를 가져와서
uid의 카테고리에 해당하는 것들을 제외한 카테고리 개수를 카운트하도록 수정

dev-msj committed:
style : 사용되지 않는 모델 삭제

UserCategoryCount
UserCategoryCountSelect

dev-msj committed:
feat(FeedMapper) : getFeedListByUidList 쿼리 추가

uid list를 받아 해당 유저들의 feed 목록을 반환

dev-msj committed:
refactor(UserRelationService) : FeedMapper를 통해 feed count하도록 변경

userCategoryCountMap으로 RecommendUser를 미리 생성.
getFeedListByUidList의 결과로 feed count하고 생성된 RecommendUser에 반영

dev-msj committed:
refactor(UserInfoMapper) : getUserFeedCountList 쿼리 변경

메서드명 변경
- getUserFeedCountList -> getDefaultRecommendUserExcludeUidList

쿼리 변경
- exclude uid list를 제외한 유저를 피드 작성량 순으로 정렬하여 30명을 가져오도록 수정

dev-msj committed:
refactor(UserInfoRepository) : UserInfoMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : UserInfoMapper 변경사항 반영

로직 정리
주석 수정

dev-msj committed:
style : 사용되지 않는 모델 삭제

UserFeedCount
UserFeedCountSelect

dev-msj committed:
style(UserCategoryMapper) : 메서드명 변경

메서드명 변경
- getOverlappingUserCategoryList -> getUserCategoryListByUidList

dev-msj committed:
style : UserCategoryMapper 수정사항 반영

dev-msj committed:
test(UserRelationServiceTest) : 서비스 로직 변경 사항 반영

dev-msj committed:
fix(UserInfoMapper) : getDefaultRecommendUserExcludeUidList 쿼리 오류 해결

outer join을 not exists로 대체하여 해결

dev-msj committed:
refactor(UserCategoryMapper) : 쿼리 최적화

불필요한 쿼리 삭제
- getUserCategoryListByUidList

메서드명 변경
- getUidListWithOverlappingBookCategory -> getUserCategoryListIfOverlappingBookCategory

getUserCategoryListIfOverlappingBookCategory 쿼리 변경
- select 대상을 자신을 제외한 유저가 아닌, exclude uid에 해당하는 유저들을 제외하도록 수정
- 카테고리가 중복되는 유저의 uid가 아닌 UserCategory를 반환하도록 수정

dev-msj committed:
refactor(UserCategoryRepository) : UserCategoryMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : UserCategoryMapper 변경사항 반영

로직 정리
30명 체크를 상수로 처리

dev-msj committed:
test(UserRelationServiceTest) : 로직 변경에 의한 테스트 코드 수정

dev-msj committed:
refactor(UserRelationRepository) : CacheEvict 삭제

dev-msj committed:
refactor(UserRelationService) : createUserRelation에 CachePut 추가

CachePut을 통해 신규 UserRelation을 redis에도 반영하도록 변경

dev-msj committed:
refactor(UserRelationController) : createUserRelation 변경

UserRelation List를 반환하도록 변경
Caching을 위해 uid를 받아 넘기도록 변경

dev-msj committed:
fix(UserRelationService) : createUserRelation 오류 해결

uid를 userRelation에서 가져오도록 수정

jinspark-lab commented on src/main/java/com/flab/oasis/controller/UserProfileController.java:
```diff
@@ -0,0 +1,34 @@
+
+    @GetMapping("/profile")
+    public UserProfileResponse getUserProfileByUid() {
+        return userProfileService.getUserProfileByUid();
```
by~ 네이밍은 파라미터 등을 기대할때 주로 사용하므로 getUserProfile 이 나을 것 같습니다

dev-msj replied:
네 해당 내용 반영했습니다

jinspark-lab commented on src/main/java/com/flab/oasis/model/response/FailureResponse.java:
```diff
@@ -0,0 +1,20 @@
+
+import com.flab.oasis.constant.ErrorCode;
+import com.flab.oasis.model.exception.UnexpectedCodeException;
+import lombok.experimental.SuperBuilder;
```
개인프로젝트라 상관은 없지만, 이런 experimental package 를 사용하실 때에는 리서치가 필요합니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserRelationService.java:
```diff
@@ -0,0 +1,120 @@
+
+        return recommendUserList.stream()
+                .map(RecommendUser::getUid)
+                .collect(Collectors.toList());
```
uidList 를 반환받은 이후부터 여기까지 stream 구문 하나로 축약할 수 있을 것 같습니다. 가능하실까요?

dev-msj replied:
네 수정했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/repository/UserCategoryRepository.java:
```diff
@@ -0,0 +1,28 @@
+        return userCategoryMapper.getUserCategoryListByUid(uid);
+    }
+
+    public List<UserCategory> getUserCategoryListIfOverlappingBookCategory(List<String> excludeUidList) {
```
지난시간에 말씀드린 구조와 연관이 있습니다. 쿼리가 무겁기 때문에 캐싱을 다루기 어렵죠. 이 부분 다시한번 피드백드립니다.

dev-msj replied:
네 일단 쿼리를 제가 의도한 대로 작성하지 못했는데, 기존에 의도했던 쿼리도 무거운 것 같네요...

그래서 쿼리를 2가지로 분리했습니다.
첫번째는 카테고리가 겹치는 유저를 추출하는 쿼리
두번째는 uid list를 받아 UserCategory를 추출하는 쿼리

그리고 exclude uid에 해당하는 유저를 제외하는 쿼리는 로직으로 대체했습니다.

jinspark-lab commented on src/test/java/com/flab/oasis/user/UserRelationServiceTest.java:
```diff
@@ -0,0 +1,124 @@
+        Assertions.assertAll(
+                () -> assertEquals(expectedUid1, uidList.get(0)),
+                () -> assertEquals(expectedUid2, uidList.get(1))
+        );
```
데이터에 Dependent 한 테스트인데, 이렇게 가는 것도 방법이지만 데이터셋을 넣고 시작하는 것도 방법입니다

jinspark-lab replied:
30 이라는 숫자보다는 분기 처리가 핵심입니다.

jinspark-lab reviewed (changes requested):
피드백드렸습니다

dev-msj committed:
refactor(UserRelationService) : makeRecommendUserListByCategory 로직 최적화

dev-msj committed:
refactor : getUserProfile로 메서드명 변경

dev-msj committed:
fix(FeedMapper) : getFeedListByUidList 누락된 Alias 추가

dev-msj committed:
refactor(UserCategoryMapper) : 쿼리를 더 작은 단위로 분리

getUidListIfOverlappingBookCategory를 통해 카테고리가 겹치는 유저 추출
getUserCategoryListByUidList를 통해 카테고리가 겹치는 유저의 카테고리 추출

dev-msj committed:
refactor(UserCategoryRepository) : UserCategoryMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : UserCategoryMapper 변경사항 반영

getUidListIfOverlappingBookCategory를 통해 카테고리가 겹치는 유저 추출
카테고리가 겹치는 유저들 중 exclude uid list에 해당하는 유저 필터링
getUserCategoryListByUidList를 통해 카테고리가 겹치는 유저의 카테고리 추출

dev-msj committed:
test(UserRelationServiceTest) : UserCategoryMapper 변경사항 반영

dev-msj committed:
style(UserRelationController) : 사용하지 않는 import 삭제

dev-msj committed:
feat(RedisConfig) : 추천 유저 config 추가

dev-msj committed:
refactor(UserRelationService) : getRecommendUserListByUid를 redis에 캐싱

dev-msj committed:
style : 메서드명 변경

getRecommendUserList -> getRecommendUserListByUid

dev-msj committed:
refactor(UserInfoMapper) : 기본 추천 유저 생성 쿼리 단순화

6개월 안에 가입했던 유저들 중 피드 작성량이 많은 순으로 추천 유저를 가져오도록 변경

dev-msj committed:
refactor : UserInfoMapper 변경사항 반영

dev-msj committed:
refactor(UserRelationService) : RecommendUserRequest 모델 추가

dev-msj committed:
refeator(UserRelationService) : Recommen User 생성 로직 변경

메서드명 변경
- getRecommendUserListByUid -> getRecommendUserList

파라메터 변경
- RecommendUserRequest를 통해 uid와 checkSize를 가져오도록 변경

dev-msj committed:
refactor(UserRelationController) : UserRelationService 변경사항 반영

dev-msj committed:
test(UserRelationServiceTest) : UserRelationService 변경사항 반영

dev-msj committed:
test(UserRelationServiceTest) : check size를 충족하는 케이스 추가

dev-msj committed:
feat(UserCategoryMapper) : getUidListByBookCategory 쿼리 추가

카테고리를 생성 또는 수정한 일자가 6개월 이내인 유저를 카테고리별로 가져오는 쿼리 작성

dev-msj committed:
feat(UserCategoryRepository) : getUidListByBookCategory 메서드 추가

redis에 메서드 실행 결과가 캐싱되도록 Cacheable 어노테이션 추가

dev-msj committed:
feat(RedisConfig) : "uidListByBookCategory"의 configuration 추가

UserCategoryRepository의 getUidListByBookCategory 메서드의 실행 결과가
1일 동안 캐싱되도록 설정

dev-msj committed:
refactor(UserRelationService) : 카테고리 추천 유저 추출 로직 변경

기존 - 쿼리로 카테고리가 겹치는 유저를 가져와 유져별 카테고리를 다시 가져오는 로직 삭제

변경 - 쿼리로 카테고리를 생성 또는 수정한 일자가 6개월 이내인 유저를 카테고리별로 가져와서
         로직으로 카테고리가 겹치는 유저를 처리

dev-msj committed:
test(UserRelationServiceTest) : UserRelationService 변경사항 반영

dev-msj committed:
style(UserCategoryMapper) : 사용하지 않는 쿼리 삭제

dev-msj committed:
style(USerCategoryRepository) : 사용되지 않는 메서드 삭제

dev-msj committed:
refactor(FeedMapper) : 피드 가져오기 쿼리 변경

uid list에 해당하는 피드 목록이 아닌 단일 대상 피드 목록을 가져오도록 변경

dev-msj committed:
refactor(UserRelationService) : 피드 개수 카운트 로직 변경

카테고리가 겹치는 유저 목록을 통해 피드를 가져오지 않고, 각 유저별로 피드를 가져오도록 변경
- 각 유저들의 피드 목록은 기본적으로 redis 캐싱 대상임

dev-msj committed:
test(UserRelationServiceTest) : UserRelationService 변경사항 적용

사용되지 않는 변수 삭제
주석 추가
주석 오류 삭제

dev-msj committed:
refactor(UserRelationService) : empty 체크 조건 변경

UserCategoryRepository의 getUserCategoryByUid의 결과가 empty일 경우
기본 추천 유저를 반환하도록 변경

dev-msj committed:
test(UserRelationSerivceTest) : UserRelationService 변경사항 적용

테스트 케이스 추가
- 설정한 카테고리가 존재하지 않을 경우

dev-msj committed:
refactor(UserRelation) : Builder 어노테이션 추가

dev-msj committed:
test(UserRelationServiceTest) : exclude user에 해당하는 케이스 추가

dev-msj committed:
feat(RedisConfig) : defaultRecommendUser의 cache configuration 추가

dev-msj committed:
refactor(UserInfoMapper) : getDefaultRecommendUser 쿼리 수정

6개월 안에 UserInfo를 생성하거나 수정한 유저를 가져오도록 수정

dev-msj committed:
refactor(UserInfoRepository) : UserInfoMapper 변경사항 반영

파라메터 삭제
Cacheable 어노테이션 추가

dev-msj committed:
refactor(UserRelationService) : UserInfoMapper 변경사항 반영

추천 유저를 가져오고, 각 추천 유저들의 피드 목록를 가져와 개수를 카운트한 후
정렬한 결과를 반환하도록 변경

주석 오류 수정

dev-msj committed:
test(UserRelationServiceTest) : UserRelationService 변경사항 반영

변수명 오류 수정
주석 오류 수정

dev-msj committed:
refactor(RecommendUser) : compareTo 메서드 수정

jinspark-lab commented on src/main/java/com/flab/oasis/model/RecommendUserRequest.java:
```diff
@@ -0,0 +1,11 @@
+@Builder
+public class RecommendUserRequest {
+    private String uid;
+    private int checkSize;
```
기획에 따라 다르지만 일반적으로 checksize 같은 값은 클라이언트가 아니라 서버에서 제어합니다.
이유를 생각해보시면 좋을 것 같습니다.

jinspark-lab commented on src/main/resources/mapper/user/UserCategoryMapper.xml:
```diff
@@ -17,22 +17,11 @@
+        SELECT UID
+        FROM USER_CATEGORY
+        WHERE BOOK_CATEGORY = #{bookCategory} AND
+              MODIFY_DATETIME BETWEEN DATE_SUB(NOW(), INTERVAL 6 MONTH) AND NOW()
```
마찬가지로 쿼리에서 BETWEEN / NOW / INTERVAL 연산을 하는 것 보다는 어플리케이션에서 하는게 좋습니다.
이유를 한번 찾아보시고, 어려우시다면 수업시간에 말씀드리겠습니다.

dev-msj replied:
서비스 운영 및 유지보수 관점에서 크게 2가지 부분에서 문제점이 예상됩니다.

첫번째로는 하나의 요청이 가지는 연산량이 많아 리소스를 많이 사용하게 될 경우입니다.
서버의 경우에는 서버를 추가하여 요청을 분산시킬 수 있습니다. 또한 클라우드 서비스에서는 Auto Scaling 기능을 활용해 실시간으로 서버의 양을 조절함으로써 훨씬 더 유연하게 대응할 수 있습니다.
하지만 DB의 경우에는 Replication 서버를 추가 개설하여 요청을 분산시킬 수는 있지만, Master 서버에 의존적이며 데이터를 복제하는 비용이 필요하므로 실시간 대응이 불가능합니다.
그렇기 때문에 DB는 CRUD 외의 불필요한 연산을 최소화하여 리소스를 절약하는 것이 좋을 것으로 생각됩니다.

두번째로는 DB에서 데이터를 필터링 하는 형태는 DB 의존성이 형성되기 때문에 유지보수 측면에서 비효율적이라고 생각됩니다.
이 경우에 요구사항 변경될 경우 쿼리와 로직 모두 변경해야 되야한다는 문제가 발생됩니다.
그래서 쿼리는 단순히 타겟 데이터의 원본을 들고 오는 형태로만 사용하여 데이터 의존성을 분리시켜줘야 한다고 생각됩니다.

dev-msj replied:
세컨더리 인덱스를 활용한 BETWEEN 연산의 경우에는 InnoDB의 구조와 관련한 이슈가 있습니다.

InnoDB는 Primary Key를 기준으로 인덱싱한 데이터 파일을 가지고 있고, 세컨더리 인덱스의 경우에는 데이터 파일 외부의 리프 노드에서 관리하고 있습니다. 그리고 각 인덱스들은 데이터 파일 상에 있는 실제 데이터들의 레코드 주소값을 가지고 있습니다. 이 때 BETWEEN 연산을 하게 되면 range 탐색 범위 안의 각각 인덱스 별로 데이터 파일에 접근하기 위한 랜덤 디스크 I/O가 발생하게 됩니다.
그래서 Primary Key를 이용해 데이터를 불러와 로직 상에서 처리하는 것이 더 효율적일 것이라 생각됩니다.

다만 추가적으로 발생하는 문제점은 현재 Primary Key를 복합키로 사용하고 있는데, UID, Book Category 순으로 되어 있다는 것입니다. Book Category만 사용해 select를 할 경우 PK가 적용되지 않고 Full Scan을 하게 됩니다.
그러면 다시 Book Category에 세컨더리 인덱스를 적용해야 하겠네요.

modify_datetime을 인덱스로 하여 range 탐색을 할 경우 range 조건에 맞는 모든 데이터를 탐색하면서 원하는 book_category를 가져오지만, book_category를 인덱스로 할 경우 book_category만 비교해서 가져오면 되니
동일하게 랜덤 디스크 I/O가 발생해도 총 비용면에서 book_category 인덱스를 설정하는 것이 더 저렴할 것으로 생각됩니다.

jinspark-lab commented on src/test/java/com/flab/oasis/user/UserRelationServiceTest.java:
```diff
@@ -80,50 +128,126 @@ void testRecommendUserSizeLessThen30() {
+                RecommendUserRequest.builder()
+                        .uid(uid)
+                        .checkSize(2)
+                        .build()
```
지난 시간에 가이드드린 방법은 getRecommendUserList 에 파라미터로 CheckSize 를 넘기는 방법이었습니다. 위의 피드백과 같이 고민해보세요.

jinspark-lab commented on src/main/java/com/flab/oasis/controller/UserRelationController.java:
```diff
@@ -17,7 +18,12 @@ public class UserRelationController {
+                RecommendUserRequest.builder()
+                        .uid(userAuthService.getAuthenticatedUid())
+                        .checkSize(30)
+                        .build()
```
보통 Request 객체는 실제 End User Request 로부터 받습니다. Controller 에서 빌딩하는 구조가 좀 어색하네요.

dev-msj replied:
지금까지 Service에 필요한 파라메터가 여러 개일 경우, 모델을 만들어 Request을 만들어서 넘겨왔기 때문에 단순하게 접근했던 것 같습니다. 이 경우 End User의 Request로부터 값을 전달받아 넘기기 때문에 의미상으로 적합한 방식이었지만, 이번 케이스의 경우에 서버 자체적으로 관리하고 있는 값을 넘기기 때문에 적합하지 않네요...
현재 Request 모델을 삭제하고 메서드 파라메터로 uid와 checkSize를 전달하도록 수정했습니다.

jinspark-lab reviewed (changes requested):
CheckSize 로직 관련해서만 정리하면 될 것 같습니다.

dev-msj committed:
refactor(UserRelationService) : 추천 유저 생성 메서드 수정

클라이언트로부터 받아야 하는 parameter가 없기 때문에 Request 모델을 사용하는 것은 부적합함
RecommendUserRequest를 풀어서 uid와 checkSize를 직접 전달하도록 변경하고, 이에 맞게 메서드명을 수정

메서드명 변경
- getRecommendUserList -> getRecommendUserListByUidAndCheckSize

파라메터 변경
- RecommendUserRequest -> uid, checkSize

변수명이 uid로 겹치는 지역 변수들은 역할에 맞는 네이밍으로 수정

dev-msj committed:
refactor : UserRelationService 변경사항 반영

dev-msj committed:
style(RecommendUserRequest) : 사용되지 않는 모델 삭제

dev-msj committed:
feat : RecommendCandidateUser 추가

dev-msj committed:
refactor(UserCategoryMapper) : 카테고리 추천 쿼리 변경

메서드명 변경
- getUidListByBookCategory -> getRecommendCandidateUserListByBookCategory

dev-msj committed:
refactor(RedisConfig) : cacheName 변경

uidListByBookCategory -> recommendCandidateUser

dev-msj committed:
refactor(UserCategoryRepository) : UserCategoryMapper 변경사항 반영

메서드명 변경
- getUidListByBookCategory -> getRecommendCandidateUserListByBookCategory

리턴타입 변경
- List<String> -> List<RecommendCandidateUser>

cacheNames 변경
- uidListByBookCategory -> recommendCandidateUser

dev-msj committed:
refactor(UserRelationService) : 추천 후보 유저 추출 조건 추가

6개월 안에 UserCategory를 수정하거나 추가한 유저를 추출하는 조건 추가

dev-msj committed:
test(UserRelationServiceTest) : UserCategoryMapper 변경사항 반영

dev-msj committed:
style : redis cache name 수정

dev-msj committed:
fix : Serializable 문제 해결

RecommendCandidateUser
- NoArgsConstructor, AllArgsConstructor 어노테이션 추가
- LocalDateTime Serializer 관련 패키지 추가
- LocalDateTime Serializer 어노테이션 추가

UserRelation
- NoArgsConstructor, AllArgsConstructor 어노테이션 추가

dev-msj committed:
fix(UserCategoryMapper) : 누락된 alias 추가

dev-msj committed:
feat : ConstantName enum class 추가

db의 ConstantName_Definition 테이블에 저장된 상수의 이름을 enum으로 관리한다.

dev-msj committed:
feat(ConstantDefinitionMapper) : int 타입의 상수를 가져오는 쿼리 작성

dev-msj committed:
feat(ehcache.xml) : defaultCache 추가

dev-msj committed:
feat : ConstantDefinitionRepository 추가

db에서 int 타입의 상수값을 가져오고 해당 상수를 ehcache에 캐싱한다.

dev-msj committed:
refactor(UserRelationService) : checkSize를 db에서 가져오도록 변경

dev-msj committed:
refactor(UserRelationController) : UserRelationService 변경사항 반영

dev-msj committed:
test(UserRelationServiceTest) : checkSize를 db에서 가져오도록 변경
